test/test_distributed.py

```diff
@@ -811,17 +821,29 @@ def forward(self, x):
                               target[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)].cuda(gpu_subset[0]),
                               loss)
 
-        for layer_gpu, layer_DDP in zip(model_gpu.modules(), model_DDP.module.modules()):
-            if isinstance(layer_gpu, nn.Linear):
-                self.assertEqual(layer_gpu.weight.grad, layer_DDP.weight.grad)
+        assert_equal_model(model_gpu, model_DDP)
 
-        # Run SGD and second iteration to shake out errors
-        optimizer.step()
-        self._test_DDP_helper(model_DDP,
-                              input_cpu,
+        # Update weights and run a second iteration to shake out errors
+        model_step(model_gpu)
+        model_step(model_DDP.module)
+
+        # Shuffle the input so that DDP input is different from the first iteration
+        input_cpu = input_cpu[torch.randperm(batch_size)]
+
+        # Run the second iteration for single gpu
+        self._test_DDP_helper(model_gpu,
+                              input_cpu.cuda(gpu_subset[0]),
                               target.cuda(gpu_subset[0]),
                               loss)
 
+        # Run the second iteration for DDP
+        self._test_DDP_helper(model_DDP,
+                              input_cpu[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)],
+                              target[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)].cuda(gpu_subset[0]),
+                              loss)
+
+        assert_equal_model(model_gpu, model_DDP)
```

Comment:
Would it be possible to fold the two iterations into a for loop? They look almost the same:
```python
for i in range(2):
    for model in [model_gpu, model_DDP]:
        self._test_DDP_helper(model, ...)
        model_step(model)
    assert_equal_model(model_gpu, model_DDP)
    input_cpu = input_cpu[torch.randperm(batch_size)]
```

Reply:
For the suggested inner loop: the input&target of two models are different since model_DDP only uses half of the input in each process. And model_step() applies to model_DDP.module instead of itself since other components will be overwritten by broadcast. I think it's good to call them directly for now until we have more models to compare?